CIP 0021 | Add Alonzo changes

gabrielKerekes opened the pull request.

gabrielKerekes committed:
CIP-0021 | Add Alonzo changes

gabrielKerekes commented on CIP-0021/README.md:
```diff
@@ -84,7 +86,9 @@ If a transaction contains a pool registration certificate, then it must not cont
-- mint entry.
+- mint entry;
+- any collateral input;
+- any required signer.
```
I'm not sure if `script_data_hash` should also be included here? Or is it just items which require witnesses?

janmazak replied:
Must be included here, too. Ledger code:
```
security_policy_t policyForSignTxScriptDataHash(const sign_tx_signingmode_t txSigningMode)
{
	switch (txSigningMode) {
	case SIGN_TX_SIGNINGMODE_ORDINARY_TX:
	case SIGN_TX_SIGNINGMODE_MULTISIG_TX:
	case SIGN_TX_SIGNINGMODE_PLUTUS_TX:
		SHOW();
		break;

	case SIGN_TX_SIGNINGMODE_POOL_REGISTRATION_OWNER:
	case SIGN_TX_SIGNINGMODE_POOL_REGISTRATION_OPERATOR:
		DENY();
		break;

	default:
		ASSERT(false);
	}
```

gabrielKerekes commented on CIP-0021/README.md:
```diff
@@ -5,7 +5,7 @@ Authors: Gabriel Kerekes <gabriel.kerekes@vacuumlabs.com>, Rafael Korbas <rafael
 Type: Standards
 Created: 2021-06-15
-Updated: 2021-11-23 multi-sig
+Updated: 2021-11-23 multi-sig 2021-11-22 Alonzo
```
Date should be updated later.

gabrielKerekes commented:
I couldn't come up with any more restrictions which aren't implementation details e.g. collateral inputs only in plutus signing mode. Suggestions are welcome.

janmazak commented on CIP-0021/README.md:
```diff
@@ -104,6 +108,10 @@ In this exceptional case, auxiliary data must be encoded in their "tuple" format
 
+**Network ID**
+
+Since Alonzo era `network_id` must always be included in the transaction body.
```
This is ambiguous, does not make it clear that network id must be included in pre-Alonzo txs as well.

janmazak commented on CIP-0021/README.md:
```diff
@@ -130,9 +138,15 @@ Similarly to multiassets, allowing duplicate withdrawals might lead to inconsist
 
+**Network ID**
+
+`network_id` is always included in the transaction body to simplify the logic of HW wallets and to simplify integration.
```
```suggestion
`network_id` is always included in the transaction body to simplify the logic of HW wallets and to simplify integration. Explicit `network_id` protects against attacks based on misleading HW wallet users about the network for which the transaction is being signed --- there are valid transactions (without outputs and withdrawals) that do not contain network id anywhere within the transaction body (and inputs are usually not shown to the user).
```

janmazak reviewed (commented):
I think we should think about restricting formally valid, but useless combinations of Alonzo elements --- don't we forbid some of them? I can't think of any.

For Plutus txs, we allow mixing stake credential types.

gabrielKerekes commented:
Suggestions applied in https://github.com/vacuumlabs/CIPs/pull/6/commits/972ecd5f820654d060c41484876cf716576fab7c.

> I think we should think about restricting formally valid, but useless combinations of Alonzo elements --- don't we forbid some of them? I can't think of any.

I don't think we forbid any combinations rather just the elements separately.

> For Plutus txs, we allow mixing stake credential types.

Hm, I removed the whole mention of stake credentials. The statement there `all the certificates included in a transaction must have the same type of stake credential i.e. either 0 - addr_keyhash or 1 - scripthash` wasn't exactly true - we allow **only** keys in ORDINARY tx, **only** scripts in MULTISIG and mixed in PLUTUS. I removed it because I'm not sure if we want to mention signing modes here as I think they are rather an implementation detail. Or if we want to mention it, here's my proposal for what we could write:
```
Stake credentials of certificates must be:
- only key hashes in ordinary (single-sig) transactions;
- only script hashes in multi-sig transactions;
- either key hashes or script hashes in Plutus transactions.
```
And the same thing for withdrawals.

janmazak commented on CIP-0021/README.md:
```diff
@@ -70,7 +72,7 @@ Unless mentioned otherwise in this CIP, optional empty lists and maps must not b
 
-Outputs containing no multi-asset tokens must be serialized as a simple tuple, i.e. `[address, coin]` instead of `[address, [coin, {}]]`.
+Outputs containing no multi-asset tokens must be serialized as a simple tuple, i.e. `[address, coin, ?datum_hash]` instead of `[address, [coin, {}], ?datum_hash]`.
 
```
`datum_hash` can only be used if the address is a script address.

janmazak reviewed (commented):
I'd add something like this to the appropriate sections:

```
Stake credentials of certificates must be:
- only key hashes in ordinary (single-sig) transactions;
- only script hashes in multi-sig transactions.
(There is no restriction for transactions involving Plutus.)

Reward accounts in withdrawals must be:
- only based on key hashes in ordinary (single-sig) transactions;
- only based on script hashes in multi-sig transactions.
(There is no restriction for transactions involving Plutus.)

```

davidmisiak commented on CIP-0021/README.md:
```diff
@@ -70,7 +72,7 @@ Unless mentioned otherwise in this CIP, optional empty lists and maps must not b
 **Outputs**
 
-Outputs containing no multi-asset tokens must be serialized as a simple tuple, i.e. `[address, coin]` instead of `[address, [coin, {}]]`.
+Outputs containing no multi-asset tokens must be serialized as a simple tuple, i.e. `[address, coin, ?datum_hash]` instead of `[address, [coin, {}], ?datum_hash]`. In addition, including `datum_hash` is only allowed if the output `address` is a script address.
```
```suggestion
Outputs containing no multi-asset tokens must be serialized as a simple tuple, i.e. `[address, coin, ?datum_hash]` instead of `[address, [coin, {}], ?datum_hash]`. In addition, including `datum_hash` is only allowed if the output `address` contains a script hash in its payment part.
```

gabrielKerekes replied:
https://github.com/vacuumlabs/CIPs/pull/6/commits/62e1226251493954b56b1f51a950c83cb365608a

davidmisiak commented on CIP-0021/README.md:
```diff
@@ -104,6 +119,10 @@ In this exceptional case, auxiliary data must be encoded in their "tuple" format
 
+**Network ID**
+
+Since Alonzo era `network_id` must be included in the transaction body of all transactions (even pre-Alonzo ones).
```
No longer required (also below).

gabrielKerekes replied:
https://github.com/vacuumlabs/CIPs/pull/6/commits/62e1226251493954b56b1f51a950c83cb365608a

janmazak commented on CIP-0021/README.md:
```diff
@@ -72,7 +71,7 @@ Unless mentioned otherwise in this CIP, optional empty lists and maps must not b
 **Outputs**
 
-Outputs containing no multi-asset tokens must be serialized as a simple tuple, i.e. `[address, coin, ?datum_hash]` instead of `[address, [coin, {}], ?datum_hash]`. In addition, including `datum_hash` is only allowed if the output `address` is a script address.
+Outputs containing no multi-asset tokens must be serialized as a simple tuple, i.e. `[address, coin, ?datum_hash]` instead of `[address, [coin, {}], ?datum_hash]`. In addition, including `datum_hash` is only allowed if the payment part of the output's `address` is a script hash.
```
We don't allow datum hash for pool registrations, it's not mentioned below.

gabrielKerekes replied:
I added to the list of forbidden items when pool registration certificate is included. https://github.com/vacuumlabs/CIPs/pull/6/commits/1c80bcf95fecfe61ba207e36c5180463797511a7 Another alternative would be to add "and the transaction doesn't contain a pool registration certificate." to this sentence, but I prefered adding it to the list.

gabrielKerekes commented:
Closing and opening a PR in parent repo.
